src/exo/API.py

```diff
@@ -267,7 +267,7 @@ def body(self):
         block = self._root()._child_block("body")
         return C.lift_cursor(block, self)
 
-    def find(self, pattern, many=False):
+    def find(self, pattern, many=False, call_depth=1):
```

Comment:
I don't think all these `call_depth` are necessary at the API level. It seems like you needed to change the value (internally) in one place and you added them to all `find` functions. I couldn't find where you are using this from user-code. I would revert most of these changes and make the change in the exact place where you needed it (within the compiler implementation).

Reply:
call_depth is necessary because we need to know how many stack frames `match_pattern` has when we get there, because we inspect the stack frame to get globals and locals. Internally, proc.find (this function) is calling into cursor.find, and both APIs are exposed to the user. More confusingly, cursor.find is calling into find (in API_cursors.py). So the problem is that multiple APIs are exposed to the user, and we have to keep track of call_depth accordingly.

Reply:
I think if you scroll through all your modifications you will find that all the uses cases where you specify `call_depth` can be statically determined and don't need to be exposed as an argument to the user. The issue is that some API functions work by calling other API functions and you need to change the `call_depth` in one of those cases from the caller to account for the extra call. You can also have some internal (private) `find` function that takes in `call_depth` and then you don't have surface this to the user code.

One other thing you could do is to refactor the logic that is inspecting the user stack here: https://github.com/exo-lang/exo/blob/91fa03bac8ddb30ea4cd4835114f1a8bea026544/src/exo/pattern_match.py#L82 and resolve at the API-level call. Then, you just pass in that information downstream instead of maintaining `call_depth` across compiler code which is somewhat confusing.

Reply:
> You can also have some internal (private) find function that takes in call_depth and then you don't have surface this to the user code.

I'm not sure what this means? Can you explain a bit more?

> refactor the logic that is inspecting the user stack here:

That would work, but then we'll end up passing around caller frameinfo across the compiler which is doubly confusing. It'll be sth like
```
    def find(self, pattern, many=False, frameinfo=None):   
        if frameinfo == None:
             frameinfo = ...                                                                                                                                                             
        return find(self._impl, self._proc, pattern, many, frameinfo=frameinfo) 
```
in `class Cursor`, which I think is worse

Also I think we do need to expose call_depth to users, because when you define a composed function in user library, you need to change the call_depth so that the stack frame points to the actual user code, not the library implementation definition.

Imagine you have `find_user_library_wrapper` for some reason in your library.
```
def find_user_library_wrapper(proc, cursor, ...):
    ... # do sth
    return proc.find(cursor, ...., call_depth=2)
```
call_depth in this case has to be 2

Reply:
With regard to the internal function, I am saying that you needed to add the `call_depth` to API-level functions because you had member API functions calling other member API functions. Instead, you can have them all call some internal non-API member function `_find` that assumes that it is always at depth = 1 because it is being called from an API function. A different way is to not make the API functions call each other.

Reply:
I see sure that'll work, but I think that'll give an incorrect stack frame for the `find_user_library_wrapper` example above

Reply:
I don't think you need that for user defined functions. You need it now because we don't have a good way to resolve patterns at the call-level, but we resolve them when we call `find` methods within the implementation. Once we add that feature this wouldn't be necessary.

Currently:
```
def user-op(proc, expr):
      e= proc.find(expr) <- resolved here
```

What we will have in the future:
```
@sched-op
def user-op(proc, expr: ExprCursor):
      # expr is already resolved before we enter the function and `call-depth` is a deterministic static value. 
```

Reply:
Yeah with `@sched-op` I believe it'll work, but that's requires more engineering and I'm not even sure what's the roadmap for that is.

Reply:
I have traced through the code and I don't think srcinfo is used in any meaningful way down this path anyways since we are not constructing an IR node. If you fail to find something, you will get something like:
```
exo.new_eff.SchedulingError: find: failed to find matches
E           Pattern: z[_] = _
```
And you can look through the python stack trace to find where the find failed. We currently don't use that information to report any such errors. I am not sure why it was important to pass the `call_depth` information to begin with.

Reply:
call_depth is used to get locals and globals in `src/exo/pattern_match.py`. Yeah we don't really good make use of srcinfo.
```
   func_locals = ChainMap(stack_frames[call_depth].frame.f_locals)
   func_globals = ChainMap(stack_frames[call_depth].frame.f_globals)
``